lib/teleterm/daemon/daemon.go

```diff
@@ -213,12 +215,78 @@ func (s *Service) createGateway(ctx context.Context, params CreateGatewayParams)
 }
 
 func (s *Service) onExpiredGatewayCert(ctx context.Context, gateway *gateway.Gateway) error {
-	cluster, err := s.ResolveCluster(gateway.TargetURI())
+	certReissuer, err := s.newGatewayCertReissuer(gateway)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	return trace.Wrap(s.cfg.GatewayCertReissuer.ReissueCert(ctx, gateway, cluster))
+	if err := certReissuer.ReissueCert(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// newGatewayCertReissuer creates a new gateway cert reissuer. If you want to
+// customize a gateway cert reissuer, for example for testing purposes, you
+// may create and configure an instance directly, without calling this method.
+func (s *Service) newGatewayCertReissuer(g *gateway.Gateway) (*gateway.CertReissuer, error) {
+	cluster, err := s.ResolveCluster(g.TargetURI())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &gateway.CertReissuer{
+		Gateway:        g,
+		ReissueDBCerts: cluster.ReissueDBCerts,
+		Relogin:        s.relogin,
+		NotifyError:    s.notifyAppAboutReissueError,
```

Comment:
nit: `s.relogin` is tied to gateway so if a relogin is required for non-gateway reasons, `Service` has to implement a new function. 

I think we can rename it to be more gateway specific. or maybe we can move creating `api.ReloginRequest` and `api.SendNotificationRequest` to the `gateway.CertReissuer`, and make these callbacks more generic.

Reply:
If it was possible, it could be completely parameterized over `api.ReloginRequest.Reason`. But with how the protobuf-generated code looks like, I don't think it's doable.

We might be left with parameterizing over `api.ReloginRequest` and `daemon.Service.relogin` handling the mutex and calling the client.

Reply:
I've replaced cert reissue with a `daemonService.retryWithRelogin`, which accomplishes the separation of concerns described here.